src/Logger.js

```diff
@@ -0,0 +1,102 @@
+const pino = require("pino");
+
+/**
+ * Wrapper class around pino package to add some extra functionality.
+ */
+module.exports = class Logger {
+    /**
+     * @param {string} loglevel  Accepted values: fatal | error | warn | info | debug | trace
+     * @param {object} context   Common props to add to all logged messages.
+     * @param {object} logger    Pino instance to use, if null create a new one (default).
+     */
+    constructor(loglevel = "info", context = {}, logger = null) {
+        this._logger = logger || pino({
+            messageKey: "message",
+            base:       null,
+            timestamp:  true,
+            level:      loglevel
+        });
+        this.setLogLevel(loglevel);
+        this.setContext(context);
+    }
+
+    /**
+     * Sets the pino-compatible log level to be used from this point forward.
+     *
+     * @param {string} loglevel  Accepted values: fatal | error | warn | info | debug | trace
+     */
+    setLogLevel(loglevel) {
+        this._logger.level = loglevel;
+    }
+
+    /**
+     * Sets common props that will be added to all logs from this point forward.
+     * Useful for things like request_id, type/namespace/service, etc.
+     *
+     * @param {object} context   Common props to add to all logged messages.
+     */
+    setContext(context) {
+        this._context = context;
+    }
+
+    /**
+     * Adds/overwrites common props that will be added to all logs from this point forward.
+     * Useful for things like request_id, type/namespace/service, etc.
+     *
+     * @param {object} context   Common props to add to all logged messages.
+     */
+    addContext(context) {
+        this._context = {...this._context, ...context};
+    }
+
+    /**
+     * Creates a child logger from the current one, inherting
+     *
+     * @param {object} context   Common props to add to all logged messages.
+     */
+    child(options) {
+        return new Logger(this._logger.level, this._context, this._logger.child(options));
+    }
+
+    info(message, props = {}) {
+        this._log("info", message, props);
+    }
+
+    warn(message, props = {}) {
+        this._log("warn", message, props);
+    }
+
+    error(message, props = {}) {
+        if (props instanceof Error) {
+            var error_props = {
+                error_message: props.message
+            };
+            if (props.hasOwnProperty("statusCode")) {
+                error_props.error_code = props.statusCode;
+            }
+            if (props.hasOwnProperty("stack")) {
+                error_props.error_stack = props.stack;
+            }
+            props = error_props;
+        }
+        this._log("error", message, props);
+    }
+
+    debug(message, props = {}) {
+        this._log("debug", message, props);
+    }
+
+    _log(loglevel, message, props) {
+        // Add loglevel and context to props for log message.
+        props = {...this._context, ...props, loglevel: loglevel.toUpperCase()};
```

Comment:
Maybe I'm a bit biased, but I like the pino approach to have immutable context. There are a couple of big pros I see: 

- you can't remove stuff set by someone else
- you can preprocess/serialize the current context to increase performances

Have you thought about preserving this approach in your design? What are the cons?

Reply:
`loglevel: loglevel.toUpperCase()`

I believe this is done automatically by logstash. I remember we have a map to translate pino numerical `level` to it's stringified version `loglevel`

Reply:
> I believe this is done automatically by logstash.
well since this is a generic package, should we keep it as is (it doesn't cause any problems) in case someone else uses it, or we later use it in another context and that is not done automatically at a later point?

>  I like the pino approach to have immutable context. [...]  Have you thought about preserving this approach in your design? What are the cons?

I ask since you were commenting this on this part of the code and not in another point: do you mean basically not allow `setContext` or `addContext` actions to change the context after logger creation? The pro is that we need this for the adserver. We instance the logger in the `utils/logger` class itself, so it's globally accessibly (the same instance) without passing it around all the time like we do in the adstitcher (would require a much bigger refactoring of the whole code to have logger always injected). And since we use the same instance everywhere and we instantiate it in the utils file, we need to be able to set the context at the beginning of the request (in the adserver currently it's done via `logger.init(event)`) so we can pass a request-wide `request_id` etc.

Reply:
> in case someone else uses it

Even the need of stringify stuff is spreaker-specific. At this stage I'd focus on our needs and ignore the external world. This is public just because there is no need for it to be private ;)

Reply:
> I ask since you were commenting this on this part of the code and not in another point

I just left the comment in one of the places in which context was used :)

> do you mean basically not allow setContext or addContext actions to change the context after logger creation?

Kind of. It was more a comment about the general approach on the use of context vs pino immutable bindings. 

> since we use the same instance everywhere and we instantiate it in the utils file, we need to be able to set the context at the beginning of the request (in the adserver currently it's done via logger.init(event)) so we can pass a request-wide request_id etc.

Ok I understand the point and the pros of the current adserver approach. Using an internal "context" vs relying a bit more on pino bindings is an implementation detail and it won't change the interface of this class anyway, it's closer to an optimization, that's something can we do it any time